### PR TITLE
PMREMGenerator: fix orientation, handedness and indexing order.

### LINKED ISSUE #19159: PMREMGenerator.fromScene does not render Scene Background correctly
##### Description of the problem

Hi there,

Currently when using an image background (in this case generated with PMREMGenerator) it seems that producing a PMREM using PREMGenerator's `fromScene` will result in a mirroring of the scene's background in the resulting Envmap. 

It's easier to show than explain, so here's an example:
https://jsfiddle.net/BruOps/e9obh3np/9/

Click the "Update PMREM" button so see what I mean. You'll see the sky flip on the models.

##### Three.js version

- [x] Dev
- [x] r115
- Since r112?? That's what I suspect at least.

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

##### Hardware Requirements (graphics card, VR Device, ...)

N/A

### PR DESCRIPTION
**Pull request by sciecode**
Fixes #19159

#### Motive:

This PR intends to fix the need for "hacky" scene scaling when using `fromScene`, however it does other things that will improve how maintainable and easy to understand the code is. Namely:

#### Changes:

- It fixes the orientation, the same way @WestLangley suggested in #19235 ( those changes are indirectly contained in this PR ).

- It also normalizes the handedness utilized in the coordinate system of the CUBE_UV texture to the one used by Three.js world coordinates, this is what actually solves the problem with the hacky scaling.

- Finally, it modifies the face indexing order to match that of the WebGL/OpenGL-ES standard. Which is the same one used on CubeTexture and CubeMaps in general. 
`faces = [ pos_X, neg_X, pos_Y, neg_Y, pos_Z, neg_Z ]`

------

#### Discussion:

The only change that is not necessary, but highly encouraged, is the changes in the indexing order. If needed, I can easily revert the last commit, the first commit is self-sufficient and only includes the changes in handedness and orientation.

As an illustration as to how this PR modifies the output texture generated by PMREMGenerator, I've run a simple example that maps how each axis is oriented in the final image.

dev ( left )   -  PR ( right )
     

------

#### Final considerations:

This PR has no impact on the final rendering, only in the cube_uv texture.

/ping @elalish @WestLangley suggestions and review are welcomed.

**requires build update**

### CODE REVIEW

**Comment by WestLangley** [comment]
Can you please recreate your before/after CubeUV images using the `pisa` cubemap?

**Comment by sciecode** [comment]
> Can you please recreate your before/after CubeUV images using the `pisa` cubemap?

Sure.


dev ( left )   -  PR ( right )

**Comment by WestLangley** [comment]
Comparing all six tiles with the original images from the cube map, +/-Y are different.

I think in the PR I filed, +Y is different from the original, and that, too, should be fixed.

**Comment by sciecode** [comment]
> Comparing all six tiles with the original images from the cube map, +/-Y are different.
> 
> I think in the PR I filed, +Y is different from the original, and that, too, should be fixed.

+/- X are also inverted, when compared to the original faces.

The reason for this is because cube maps handedness and orientation are different from that of Three.js.

Three.js ( right-handed ):

> Z+ - back 
> X+ - right
> Y+ - up ( looking up from the Z- axis )

Cube maps ( left-handed )

> Z+ - back 
> X+ - left
> Y+ - up ( looking up from the Z+ axis )

If we absolutely want to use the same orientation and handedness from cube maps, we will need to apply a scaling to the scene and therefore, look for a different solution for fixing the problems of the background exposed by #19159.

**Comment by elalish** [comment]
@sciecode I'm having a hard time groking the meaning of those axes. I can understand the camera looking along the Z axis and seeing X+ to the right (or left) and Y+ as up, but when you say Z+ is "back", does that mean the same direction the camera is pointed (toward the back of the object) or the opposite (toward the back of the camera)?

Anyway, your change looks good to me so far, modulo picking which coordinate system we want to be consistent with.

**Comment by sciecode** [comment]
I could've done a better job explaining, my bad. I was a bit tired after spending 4 hours turning my hand one way or the other trying to figure out the correct orientations of each face 😁 

**Three.js**

On Three.js we usually set the camera pointing down on the Z- axis, in this setup. X+ is to the right of the camera, Y+ is up and Z- is front; regular RH coordinate system.

As far as the images orientation, for X and Z images, we keep the `camera.up` pointing to Y+. When we want to capture +/- Y we move the camera up or down from "standard" camera position ( Z- direction ). So for Y+, camera.up becomes Z+. For Y- camera.up becomes Z-.

**CubeMap**

If we keep the camera the same way we do with three.js, the only difference is that what is X+ in three.js will be X- for cube maps. So, LF coordinate system.

However, that's not how cube map position their camera regularly. Instead they point the camera towards Z+. In that scenario, X+ is right, Y+ is up and Z+ is front.  ( in their LF CS )

For the orientation of the images, `camera.up` points to Y+ for X and Z images. For +/- Y, we move camera up or down from the "standard" camera position ( Z+ direction ). So for Y+, camera.up becomes Z-. For Y- camera.up becomes Z+.

**TL/DR**

When comparing the orientation of Three.js axis images and cube map's axis images. +/- Z are identical, +/- are swapped and +/- Y are rotated by PI.  We can rotate the "default" behavior to be whatever axis we want, however the will always be at least one axis where the images are swapped, because of RH / LH coordinate system differences.  I picked the default Three.js orientation, so we don't need to neither rotate nor reflect the scene.

**Comment by elalish** [comment]
@sciecode Thanks for the very clear explanation! I hope you can see why I got this wrong the first time through... @WestLangley this seems like a good approach to me; do you agree?

**Comment by sciecode** [comment]
I just noticed I forgot to change the indexing scheme on the `getUV` and `getDirection` methods.
If we choose to really alter the indexing, I should probably fix that to keep it consistent. If we don't, I can simply revert the last PR and everything will be already consistent.

**Comment by sciecode** [comment]
Ok, I would like to review a wrong statement I said.

Everything I just said about the coordinate system and orientation is correct, however the need for `rotation` and `scaling` is NOT!

We may in fact use the cube maps orientation and coordinate system in the cube_uv texture and just change the camera on `fromScene` method to the way I just described in the explanation post. Without the need for rotation or scaling, I just tested it and it works perfectly.

In order words, we can favor https://github.com/mrdoob/three.js/pull/19235 over this PR and just fix the camera inside `fromScene` in a future PR. ( In that case, it does make sense to change the indexing scheme as well ).

@WestLangley I think you prefer things that way right?

**Comment by WestLangley** [comment]
You are using different face indexing methods in different parts of the code. I think you are trying to do too much.

https://github.com/mrdoob/three.js/blob/aeb807ef589e2d2291217abff4a9ad6e87057de4/src/renderers/shaders/ShaderChunk/cube_uv_reflection_fragment.glsl.js#L34-L40

https://github.com/mrdoob/three.js/blob/aeb807ef589e2d2291217abff4a9ad6e87057de4/src/extras/PMREMGenerator.js#L800-L808

**Comment by WestLangley** [comment]
> I picked the default Three.js orientation, so we don't need to neither rotate nor reflect the scene.

The orientation of the three.js camera when the camera's rotation is zero has nothing to do with this issue, IMO. 

This issue is about (1) a left-hand verses right-hand coordinate system, and (2) the convention for indexing the 6 faces of a cube map.

PMREM does not use the WebGL conventions in either case.

**Comment by WestLangley** [comment]
PMREM from #19235 -- matches the 6 cubemap images
WestLangley


PMREM from this PR -- posY and negY are flipped, both horizontally and vertically
sciecode

**Comment by sciecode** [comment]
> You are using different face indexing methods in different parts of the code. I think you are trying to do too much.

Both statements are true. I acknowledged this on a previous post that I deleted. I'm in favor of dropping the indexing changes and revert to only orientation and handedness.  

However, if you are able to address the concerns exposed by https://github.com/mrdoob/three.js/pull/19235#issuecomment-626456075, I will just close this.

**Comment by sciecode** [comment]
Closing in favor of #19235.
